File: Tree/Ruby/BSTree.rb
require_relative "BSTNode"


class BSTree
	def initialize(root=nil)
		@root = root
	end # end constructor

	attr_writer :root

	def insert_node(node, elem)
		if node.nil? then 
			return BSTNode.new(elem=elem) 
		end # end if

		if elem < node.elem then
			node.left = self.insert_node(node.left, elem)
		else
			node.right = self.insert_node(node.right, elem)
		end # end if
		return node
	end # end func

	def insert(elem)
		if @root.nil? then 
			@root = BSTNode.new(elem=elem) 
		end # end if

		if elem < @root.elem then
			@root.left = self.insert_node(@root.left, elem)
		else
			@root.right = self.insert_node(@root.right, elem)
		end # end if
	end # end func

	def find_min_node(node)
		curr_node = node
		# Loop until the leftmost leaf
		while !curr_node.left.nil?
			curr_node = curr_node.left
		end # end while
		return curr_node
	end # end func

	def deleteNode(node, val)
		if node.nil? then return nil end # end if

		if node.elem == val
			then
				# When node has one or no child
				if node.left.nil?
					then return node.right
				elsif node.left.nil?
					then return node.left
				end # end inner if

				temp = self.find_min_node(node.right)
				node.elem = temp.elem
				node.right = self.deleteNode(node.right, temp.elem)
				return node
		end # end outer if

		if val < node.elem
			then node.left = self.deleteNode(node.left, val)
		else
			node.right = self.deleteNode(node.right, val)
		end # end if
		return node
	end # end func

	def delete(val)
		if @root.nil? then return nil end # end if

		if @root.elem == val
			then 
				if @root.left.nil?
					then @root = @root.right
					return
				elsif @root.right.nil?
					then @root = @root.left
					return
				else
					temp = self.find_min_node(@root.right)
					@root.elem = temp.elem
					@root.right = self.deleteNode(@root.right, temp.elem)
					return
				end # end inner if
		end # end outer if

		if val < @root.elem
			then @root.left = self.deleteNode(@root.left, val)
		else
			@root.right = self.deleteNode(@root.right, val)
		end # end if
	end # end func

	def findNode(node, val)
		if node.nil? then return false end # end if

		# If Node is null, code will crush
		if node.elem == val then return true
		elsif val < node.elem
			then return self.findNode(node.left, val)
		elsif val > node.elem
			then return self.findNode(node.right, val)
		end # end if
	end # end func

	def find(val)
		return self.findNone(@root, val)
	end # end func

	def printElem_inorder(node)
		if node.nil? then return end # end if
		self.printElem_inorder(node.left)
		print "#{node.elem} "
		self.printElem_inorder(node.right)
	end # end func

	def printElem_preorder(node)
		if node.nil? then return end # end if
		print "#{node.elem} "
		self.printElem_preorder(node.left)
		self.printElem_preorder(node.right)
	end # end func

	def printElem_postorder(node)
		if node.nil? then return end # end if
		self.printElem_postorder(node.left)
		self.printElem_postorder(node.right)
		print "#{node.elem} "
	end # end func

	def printElem(order='in')
		if order == 'in'
			then self.printElem_inorder(@root)
		elsif order == 'pre'
			then self.printElem_preorder(@root)
		elsif order == 'post'
			then self.printElem_postorder(@root)
		end # end if
		puts
	end # end func

end # end class

File: Tree/Java/BSTree.java

public class BSTree {
	protected class BSTNode{
		private Comparable data;
		private BSTNode left;
		private BSTNode right;
		
		public BSTNode(Comparable data){
			this.data = data;
		}
		
		public String toString(){
			return (String) data;
		}
	}
	
	private BSTNode root;
	
	public boolean find(Comparable value){
		return find(value,root);
	}
	
	public boolean find(Comparable value,BSTNode node){
		if (node == null)
			return false;
		if (node.data.compareTo(value)==0)
			return true;
		else if (node.data.compareTo(value)<0)
			return find(value,node.left);
		else 
			return find(value,node.right); 
	}
	
	public BSTNode insert(Comparable value){
		if (root == null){
			root = insert(value,root);
			return root;
		}
		return insert(value,root);
	}
	
	public BSTNode insert(Comparable value, BSTNode node){
		if (node == null)
			return new BSTNode(value);
		if(node.data.compareTo(value)<0){
			node.left = insert(value, node.left);
			return node;
		}
		else{
			node.right = insert(value, node.right);
			return node;
		}	
	}
	
	public BSTNode delete(Comparable value){
		root = delete(value,root);
		return root;
	}
	
	public BSTNode delete(Comparable value,BSTNode node){
		if (node == null)
			return null;
		if (node.data.compareTo(value)==0){
			if (node.left == null)
				return node.right;
			else if (node.right == null)
				return node.left;
			else{
				if(node.right.left == null){
					node.data = node.right.data;
					node.right = node.right.right;
					return node;
				}
				else{
					node.data = removeSmallest(node.right);
					return node;
				}
			}// End outer-else
		} // End outer-if
		else if (value.compareTo(node.data)<0){
			node.left = delete(value, node.left);
			return node;
		}
		else{
			node.right = delete(value, node.right);
			return node;
		}
	}// End func
	
	private Comparable removeSmallest(BSTNode node){
		if (node.left.left == null){
			Comparable smallest = node.left.data;
			node.left = node.left.right;
			return smallest;
		}
		else
			return removeSmallest(node.left);
			
	}
	
	
	private String print(BSTNode node){
		if (node == null)
			return "";
		String str = "";
		str += print(node.right);
		str += node.data+",";
		str += print(node.left);
		return str;
	}
	
	public String toStringInOrder(){
		String str = print(root);
		String[] elem = str.split(",");
		str = "";
		for(int i=0;i<elem.length;i++){
			str += elem[i];
			if(i != elem.length-1)
				str += " ";
		}
		return str;
	}
	
	public String printPreOrder(BSTNode node) {
		if (node == null)
			return "";
		String str = "";
		str += node.data +",";
		str += print(node.left);
		str += print(node.right);
		return str;
	}
	public String toStringPreOrder(){
		String str = printPreOrder(root);
		String[] elem = str.split(",");
		str = "";
		for(int i=0;i<elem.length;i++){
			str += elem[i];
			if(i != elem.length-1)
				str += " ";
		}
		return str;
	}
	
}
